service/history/shard/controller_impl.go

```diff
@@ -75,6 +75,14 @@ type (
 		ownership            *ownership
 		status               int32
 		taggedMetricsHandler metrics.Handler
+		// shardCountSubscriptions is a set of subscriptions that receive shard count updates whenever the set of
+		// shards that this controller owns changes.
+		shardCountSubscriptions map[*shardCountSubscription]struct{}
+	}
+	// shardCountSubscription is a subscription to shard count updates.
+	shardCountSubscription struct {
+		controller *ControllerImpl
```

Comment:
nit: not sure if using a callback will be better here instead of referencing and mutating controller state directly, but no strong opinion.